src/audio/tdfb/tdfb_ipc4.c

```diff
@@ -0,0 +1,198 @@
+#include <rtos/string.h>
+#include <ipc/stream.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/data_blob.h>
+#include <sof/list.h>
+#include <stdint.h>
+
+#include "tdfb.h"
+#include "tdfb_comp.h"
+
+struct tdfb_notification_payload {
+	struct sof_ipc4_notify_module_data module_data;
+	struct sof_ipc4_control_msg_payload control_msg;
+	struct sof_ipc4_ctrl_value_chan control_value; /* One channel value */
+};
+
+LOG_MODULE_DECLARE(tdfb, CONFIG_SOF_LOG_LEVEL);
+
+static struct ipc_msg *tdfb_notification_init(struct processing_module *mod,
+					      uint32_t control_type_param_id,
+					      uint32_t control_id)
+{
+	struct ipc_msg msg_proto;
+	struct comp_dev *dev = mod->dev;
+	struct comp_ipc_config *ipc_config = &dev->ipc_config;
+	union ipc4_notification_header *primary =
+		(union ipc4_notification_header *)&msg_proto.header;
+	struct ipc_msg *msg;
+	struct tdfb_notification_payload *payload;
+
+	/* Clear header, extension, and other ipc_msg members */
+	memset_s(&msg_proto, sizeof(msg_proto), 0, sizeof(msg_proto));
+	primary->r.notif_type = SOF_IPC4_MODULE_NOTIFICATION;
+	primary->r.type = SOF_IPC4_GLB_NOTIFICATION;
+	primary->r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REQUEST;
+	primary->r.msg_tgt = SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG;
+	msg = ipc_msg_w_ext_init(msg_proto.header, msg_proto.extension,
+				 sizeof(struct tdfb_notification_payload));
+	if (!msg)
+		return NULL;
+
+	payload = (struct tdfb_notification_payload *)msg->tx_data;
+	payload->module_data.instance_id = IPC4_INST_ID(ipc_config->id);
+	payload->module_data.module_id = IPC4_MOD_ID(ipc_config->id);
+	payload->module_data.event_id = SOF_IPC4_NOTIFY_MODULE_EVENTID_ALSA_MAGIC_VAL |
+		control_type_param_id;
+	payload->module_data.event_data_size = sizeof(struct sof_ipc4_control_msg_payload) +
+		sizeof(struct sof_ipc4_ctrl_value_chan);
+	payload->control_msg.id = control_id;
+	payload->control_msg.num_elems = 1;
+	payload->control_value.channel = 0;
+
+	comp_dbg(dev, "instance_id = 0x%08x, module_id = 0x%08x",
+		 payload->module_data.instance_id, payload->module_data.module_id);
+	return msg;
+}
+
+static void tdfb_send_notification(struct ipc_msg *msg, uint32_t val)
+{
+	struct tdfb_notification_payload *ipc_payload;
+
+	ipc_payload = (struct tdfb_notification_payload *)msg->tx_data;
+	ipc_payload->control_value.value = val;
+	ipc_msg_send(msg, NULL, false);
+}
+
+int tdfb_common_notification_init(struct processing_module *mod)
+{
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	cd->msg = tdfb_notification_init(mod, SOF_IPC4_ENUM_CONTROL_PARAM_ID,
+					 CTRL_INDEX_AZIMUTH_ESTIMATE);
+	if (!cd->msg) {
+		comp_err(mod->dev, "Failed to initialize control notification.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+void tdfb_send_notification_common(struct processing_module *mod)
+{
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	tdfb_send_notification(cd->msg, cd->az_value_estimate);
+}
+
+int tdfb_get_config_common(struct processing_module *mod,
```

Comment:
sorry, not sure why you call these functions `*_common()`? They differ between IPC3 and IPC4, I would use "common" for functions shared between IPC versions?

Reply:
ack, sounds like they are not common, otherwise it LGTM. Lets fix the ipc/common naming.

Reply:
both ipc3/ipc4 have this function implementation, my purpose is:
keep interface function defined in main source file, i.e, keep: tdfb_get_config & tdfb_set_config in tdfb.c.

Otherwise, I can move tdfb_get_config & tdfb_get_config to ipc3/ipc4 separately, this will make tdfb.c looks lack something

Reply:
the code separation is good, we're only proposing to change the function naming. "common" in this case suggests that those functions are *common* for IPC3 and IPC4. The code in tdfb.c is common for IPC3 and IPC4. But functions, implemented separately for IPC3 and IPC4 are IPC-version-specific. I'm not sure I can propose a good name suffix for those version-specific function parts, how about `..._private` or `..._priv` or `..._specific`?

Reply:
no perfect name actually, let's keep for now?

Reply:
> no perfect name actually, let's keep for now?

personally each of the proposed options seems better to me than "common," e.g. I'd go with "priv"

Reply:
@btian1 @lyakh Hmm, I agree. "common" is very misleading as the functions actually are different and not common. Naming is hard, let's go with "priv", not great, but good enough.